src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -201,14 +202,42 @@ internal static IEnumerable<string> GetCommaSeparatedValues(string text)
             return validName;
         }
 
-        private static IEnumerable<TagHelperAttributeDescriptor> GetAttributeDescriptors(Type type)
+        private static IEnumerable<TagHelperAttributeDescriptor> GetAttributeDescriptors(
+            Type type, 
+            ErrorSink errorSink)
         {
-            var properties = type.GetRuntimeProperties().Where(IsValidProperty);
+            var properties = type.GetRuntimeProperties().Where(IsAccessibleProperty);
             var attributeDescriptors = properties.Select(ToAttributeDescriptor);
 
+            attributeDescriptors = attributeDescriptors.Where(
+                descriptor => ValidTagHelperAttributeDescriptor(descriptor, type, errorSink));
+
             return attributeDescriptors;
         }
 
+        private static bool ValidTagHelperAttributeDescriptor(
```

Comment:
still nesting state machines.  how about
``` c#
foreach (var property in type.GetRuntimeProperties())
{
  if (!IsValidProperty())
  {
    continue;
  }

  var descriptor ...
}
```

I'm also not excited about creating the `TagHelperAttributeDescriptor` only to throw it away if the name turns out to start with `data-`.  probably better to calculate the name, validate it, and pass it to `ToAttributeDescriptor()` (along with the `PropertyInfo`).